rest-service/manager_rest/cluster_status_manager.py

```diff
@@ -0,0 +1,240 @@
+        service_statuses = {node['status'] for name, node in
+                            formatted_nodes.items()}
+
+        # If there is only one type of status it is it, else Degraded
+        if len(service_statuses) == 1:
+            return service_statuses.pop()
+        return ServiceStatus.DEGRADED
+
+    def _is_report_updated(self, node_status):
+        """
+        Time since reported should be maximum twice the frequency interval
+        (Nyquist sampling theorem)
+        """
+        reporting_delta = node_status['reporting_freq'] * 2
+        report_time = parse_datetime_string(node_status['timestamp'])
+        return (report_time + timedelta(seconds=reporting_delta) >
+                datetime.utcnow())
+
+    def _keys_in_dict(self, dictionary, keys):
+        return all(key in dictionary for key in keys)
+
+    def _is_report_content_valid(self, node_status):
+        return (
+            self._keys_in_dict(node_status, ['reporting_freq', 'report',
+                                             'timestamp'])
+            and self._keys_in_dict(node_status['report'], ['status',
+                                                           'services'])
+            and node_status['report']['status'] in [ServiceStatus.HEALTHY,
+                                                    ServiceStatus.FAIL]
+        )
+
+    def _is_report_valid(self, node_status):
+        return (self._is_report_updated(node_status) and
+                self._is_report_content_valid(node_status))
+
+    def _handle_missing_status_files(self):
+        """Add status data to the nodes with missing status files"""
+        if not self.missing_status_files:
+            return
+
+        manager_service = self.cluster_status['services'][
+            CloudifyNodeType.MANAGER]
+        self._handle_missing_manager_files(manager_service)
+
+        services_types = [(CloudifyNodeType.DB, DB_SERVICE_KEY),
+                          (CloudifyNodeType.BROKER, BROKER_SERVICE_KEY)]
+        for service_type, service_key in services_types:
+            self._handle_missing_non_manager_files(service_type,
+                                                   service_key,
+                                                   manager_service['nodes'])
+
+    def _handle_missing_manager_files(self, manager_service):
+        missing_managers = self.missing_status_files.get(
+            CloudifyNodeType.MANAGER)
+        if not missing_managers:
+            return
+
+        manager_nodes = manager_service['nodes']
+        for manager in missing_managers:
+            manager_nodes[manager.hostname]['status'] = ServiceStatus.FAIL
+        manager_service['status'] = self._get_service_status(manager_nodes)
+
+    def _handle_missing_non_manager_files(self, service_type, service_key,
+                                          managers):
+        """
+        Missing status file because it is an external service (user-provided)
```

Comment:
=>A node's status file could be missing, due to it being an external service (user-provided)
        or all-in-one manager or it's status reporter didn't send a report ey